src/ros/node.cpp

```diff
@@ -0,0 +1,985 @@
+      saveCamConfig(node_parameters_.ids_configuration_filename);
+      node_parameters_.export_ids_configuration = true; // flag to inform the loop to unset the parameter
+      RCLCPP_INFO(this->get_logger(), "exported IDS configuration to '%s'", node_parameters_.ids_configuration_filename.c_str());
+    } catch (const std::runtime_error& e) {
+      RCLCPP_ERROR(this->get_logger(), "%s [hint: has the directory been created?]", e.what());
+      result.successful = false;
+      return result;
+    }
+  } else {
+    // The frame grabbing loop will have reverted the setting, nothing to do
+  }
+
+  // Don't print the updated configuration if it's just the export_ids_configuration flag setting/resetting
+  if ( (! changed_node_parameters.empty()) || (!changed_parameters.empty()) ) {
+    printConfiguration(); // debugging
+  }
+  return result;
+}
+
+/*****************************************************************************
+ ** Intrinsics
+ *****************************************************************************/
+
+void Node::loadIntrinsicsFile() {
+  if (camera_calibration_parsers::readCalibration(node_parameters_.camera_intrinsics_filename, node_parameters_.camera_name, ros_cam_info_)) {
+    RCLCPP_DEBUG(this->get_logger(), "Loaded intrinsics parameters for [%s]", node_parameters_.camera_name.c_str());
+  }
+}
+
+bool Node::saveIntrinsicsFile() {
+  if (camera_calibration_parsers::writeCalibration(node_parameters_.camera_intrinsics_filename, node_parameters_.camera_name, ros_cam_info_)) {
+    RCLCPP_DEBUG(this->get_logger(), "Saved intrinsics parameters for [%s] to %s", node_parameters_.camera_name.c_str(), node_parameters_.camera_intrinsics_filename.c_str());
+    return true;
+  }
+  return false;
+}
+
+/*****************************************************************************
+ ** TimeStamping
+ *****************************************************************************/
+
+rclcpp::Time Node::getImageTimestamp() {
+  // There have been several issues reported on time drifts, so we shall rely purely on ros::Time::now()
+  // e.g. https://github.com/anqixu/ueye_cam/issues/82
+  //
+  // UEYETIME utime;
+  // if(getTimestamp(&utime)) {
+  //   struct tm tm;
+  //   tm.tm_year = utime.wYear - 1900;
+  //   tm.tm_mon = utime.wMonth - 1;
+  //   tm.tm_mday = utime.wDay;
+  //   tm.tm_hour = utime.wHour;
+  //   tm.tm_min = utime.wMinute;
+  //   tm.tm_sec = utime.wSecond;
+  //   ros::Time t = ros::Time(mktime(&tm),utime.wMilliseconds*1e6);
+  //
+  //   // Deal with instability due to daylight savings time
+  //   if (abs((ros::Time::now() - t).toSec()) > abs((ros::Time::now() - (t+ros::Duration(3600,0))).toSec())) { t += ros::Duration(3600,0); }
+  //   if (abs((ros::Time::now() - t).toSec()) > abs((ros::Time::now() - (t-ros::Duration(3600,0))).toSec())) { t -= ros::Duration(3600,0); }
+  //
+  //   return t;
+  // }
+  return this->now();
+}
+
+rclcpp::Time Node::getImageTickTimestamp() {
```

Comment:
I recall that there have been several reported issues of time drift with a similar approach, resulting in the ROS1 version abandoning this approach and just returning the current ROS time for stamping the image.

Reply:
Aye, didn't try to debug these problems with this PR.

Reply:
Yeah, we can keep it as is for now since this is ideally how the timestamping should be done. If any drift is seen, we may have to revert back to simply returning the ros time. Accurate timestamping is pretty important for stuff like state estimation, so I just want to be sensitive to this issue.